### DISCUSSION TITLE
Unable to set "special" environment variable

### DISCUSSION THREAD

**Discussion by neilyoung** [community/community#26635] [Actions]
<p>A gist requires me to set <strong>DYLD_FALLBACK_LIBRARY_PATH</strong> for a Rust compilation. I tried everything, but I’m completely unable to set this env variable. It is always empty. I tried a different name, that works… Just why?</p>
<pre><code class="lang-yaml">name: Test

on:
  workflow_dispatch

jobs:
  test_job:
    runs-on: macos-latest
    steps:
      - name: Test
        run: |
          echo "PATHX=$(xcode-select --print-path)/Toolchains/XcodeDefault.xctoolchain/usr/lib/" &gt;&gt; $GITHUB_ENV
          
      - name: Test
        run: |
          echo "DYLD_FALLBACK_LIBRARY_PATH=$PATHX" &gt;&gt; $GITHUB_ENV
      

      - name: Test
        run: |
          echo ${PATHX:-is_empty}
          echo ${DYLD_FALLBACK_LIBRARY_PATH:-is_empty}
         
</code></pre>
<p>Output:</p>
<pre><code class="lang-bash">Run echo ${PATHX:-is_empty}
  echo ${PATHX:-is_empty}
  echo ${DYLD_FALLBACK_LIBRARY_PATH:-is_empty}
  shell: /bin/bash -e {0}
  env:
    PATHX: /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/
    DYLD_FALLBACK_LIBRARY_PATH: /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/
/Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/
is_empty
</code></pre>

**Comment by neilyoung** [comment]
<p>Looks like a System Integrity Protection issue</p>

**Comment by jsoref** [comment, marked as answer]
<p>It’s possible to work around SIP, you just need binaries that aren’t system binaries and that don’t live in system paths.</p>
<p>Here’s an example using <code>brew</code> to install <code>bash</code> and then using that <code>bash</code>:</p><aside class="onebox githubactions" data-onebox-src="https://github.com/jsoref/macos-sip/actions/runs/2385387464/workflow">
  <header class="source">

      <a href="https://github.com/jsoref/macos-sip/actions/runs/2385387464/workflow" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
      <svg width="24" height="24" class="github-icon github-icon-success" viewbox="0 0 24 24" aria-hidden="true"><path d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12zm16.28-2.72a.75.75 0 00-1.06-1.06l-5.97 5.97-2.47-2.47a.75.75 0 00-1.06 1.06l3 3a.75.75 0 001.06 0l6.5-6.5z"></path></svg>



  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/jsoref/macos-sip/actions/runs/2385387464/workflow" target="_blank" rel="noopener">Update test.yml</a>
    </h4>

      <div class="github-info">
        Test <span class="github-run-number">#6</span>
      </div>
  </div>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Note: I didn’t fill in <code>$PATHX</code>, but that’s not a big deal</p>

**Comment by neilyoung** [comment]
<p>Very nice. Indeed this works perfectly. I suppose I will then always have to use the shell installed by brew, right (as shown)?</p>
<p>Updated script (the PATHX was just a test if it works with non-suspect variable names):</p>
<pre><code class="lang-yaml">
name: Test

on:
  workflow_dispatch

jobs:
  test_job:
    runs-on: macos-latest
    steps:
      - name: brew bash
        run: brew install bash

      - name: Test (shim env)
        shell: /usr/local/bin/bash {0}
        run: |
          echo "DYLD_FALLBACK_LIBRARY_PATH=$(xcode-select --print-path)/Toolchains/XcodeDefault.xctoolchain/usr/lib/" &gt;&gt; $GITHUB_ENV
      

      - name: Test
        shell: /usr/local/bin/bash {0}
        run: |
          echo ${DYLD_FALLBACK_LIBRARY_PATH:-is_empty}
</code></pre>
<p>Output:</p>
<p>“/Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/”</p>

**Comment by neilyoung** [comment]
<p>This even works for the desired use case.</p>
<p>Unfortunately the created binary does run on Intel hardware only, since it has external references to OpenCV4 installed by brew. Brew uses different paths on Intel and M1 (Intel <code>/usr/local/opt/opencv/lib/</code>, M1 <code>/opt/homebrew/Cellar/opencv/4.5.5_2/lib/</code>). This makes the app work fine running on Intel hardware, but failing on M1 with</p>
<pre><code class="lang-auto">~/Downloads $ ./calibration
dyld[34291]: Library not loaded: /usr/local/opt/opencv/lib/libopencv_gapi.405.dylib
  Referenced from: /Users/decades/Downloads/calibration
  Reason: tried: '/usr/local/opt/opencv/lib/libopencv_gapi.405.dylib' (no such file), '/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib//libopencv_gapi.405.dylib' (no such file)
Abort trap: 6
</code></pre>
<p>But this is another story :slight_smile:</p>
<p>Thanks for now.</p>
